service/session.cc

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "service/session.hh"
+#include "log.hh"
+
+namespace service {
+
+static logging::logger slogger("session");
+
+static thread_local session_manager the_session_manager;
+
+session_manager& get_session_manager() {
+    return the_session_manager;
+}
+
+session_manager::session_manager() {
+    create_session(default_session_id);
+}
+
+session::guard session_manager::enter_session(session_id id) {
+    auto i = sessions.find(id);
+    if (i == sessions.end()) {
+        throw std::runtime_error(fmt::format("Session not found: {}", id));
+    }
+    auto guard = i->second->enter();
+    slogger.debug("session {} entered", id);
+    return guard;
+}
+
+void session_manager::create_session(session_id id) {
+    auto [i, created] = sessions.emplace(id, std::make_unique<session>(id));
+    if (created) {
+        slogger.debug("session {} created", id);
+    } else {
+        slogger.debug("session {} already exists", id);
+    }
+}
+
+void session_manager::prune_sessions_except(const std::unordered_set<session_id>& keep) {
+    for (auto&& [id, session] : sessions) {
+        if (!keep.contains(id)) {
+            session->start_closing();
+        }
+    }
+}
+
+future<> session_manager::drain_closing_sessions() {
+    auto lock = co_await get_units(session_drain_sem, 1);
+    auto i = closing_sessions.begin();
```

Comment:
Where do we insert the closing sessions  into closing_sessions? I do not see it.

Reply:
Right, missed that.